src/components/flowchart/flowchart.js

```diff
@@ -528,34 +538,80 @@ export class FlowChart extends Component {
     // then on a single node click, it should reset the sliced pipeline state
     if (from !== null && to !== null) {
       this.updateSlicedPipelineState(null, null, []);
+      this.setState({ showSlicingNotification: false }); // Hide notification
     } else {
       // Else, set the first node as the 'from' node based on current state
       // we need this so that if user hold shift and click on a second node,
       // the 'from' node is already set
       this.updateSlicedPipelineState(id, to, range);
+      this.setState({ showSlicingNotification: true }); // Show notification
     }
 
     // Reset the slicePipeline on single node click
     this.props.onSlicePipeline(null, null);
   };
 
+  /**
+   * Determines the correct order of nodes based on their positions,
+   * but keeps the user's original selection as the visual 'from' node.
+   * @param {string} userSelectedFromNodeId - User's initially selected 'from' node ID.
+   * @param {string} toNodeId - The current 'to' node ID.
+   * @returns {Object} - Object containing userVisualFromNodeId and adjustedFromNodeId, newToNodeId
+   */
+  determineNodesOrder = (userSelectedFromNodeId, toNodeId) => {
+    // Get bounding client rects of nodes
+    const fromNodeElement = document.querySelector(
+      `[data-id="${userSelectedFromNodeId}"]`
+    );
+    const toNodeElement = document.querySelector(`[data-id="${toNodeId}"]`);
+
+    if (!fromNodeElement || !toNodeElement) {
+      return {
+        userVisualFromNodeId: null,
+        adjustedFromNodeId: null,
+        newToNodeId: null,
+      }; // If any element is missing, return nulls
+    }
+
+    const fromNodeRect = fromNodeElement.getBoundingClientRect();
+    const toNodeRect = toNodeElement.getBoundingClientRect();
+
+    // Reorder the nodes based on their Y-coordinate
+    const [adjustedFromNodeId, newToNodeId] =
+      fromNodeRect.y < toNodeRect.y
+        ? [userSelectedFromNodeId, toNodeId] // Keep order
+        : [toNodeId, userSelectedFromNodeId]; // Swap if needed
+
+    return {
+      userVisualFromNodeId: userSelectedFromNodeId, // Keep user's selection visually as 'from'
```

Comment:
i don't think you need to return this to keep the users selection. Different approach as the comment below

Reply:
also i took me awhile to fully understand what these `adjustedFromNodeId and newToNodeId` mean. Maybe something simpler might be better, something to indicated that this `fromNodeId` and `toNodeId` has been updated/re-ordered their position?